src/reaction_network.jl

```diff
@@ -562,27 +577,89 @@ function calculate_symjac(reactions, reactants, parameters, syms)
                 j = reactants[dep]
 
                 @inbounds for ns in rx.netstoich
-                    i = reactants[ns.reactant]
-                    jacexprs[i,j] = addstoich_to_exprsum(jacexprs[i,j], dratelaw, ns.stoichiometry)
+                    key = resolveindex(getkeytype(jacexprs), reactants[ns.reactant], j)
+                    if checkforkey(jacexprs, key) 
+                        jacexprs[key] = addstoich_to_exprsum(jacexprs[key], dratelaw, ns.stoichiometry)
+                    else
+                        jacexprs[key] = addstoich_to_exprsum(0, dratelaw, ns.stoichiometry)
+                    end
                 end
             end
         end
     end
+    nothing
+end
 
+# generate a dense matrix of ExprValues representing the Jacobian
+function calculate_symjac(reactions, reactants)
+    nspecs   = length(reactants) 
+    jacexprs = Matrix{ExprValues}(undef, nspecs, nspecs)
+    fill!(jacexprs, 0)
+    jac_as_exprvalues(jacexprs, reactions, reactants)
     jacexprs
 end
 
-#Makes the Jacobian.
-function calculate_jac(symjac::Matrix{ExprValues}, reactants::OrderedDict{Symbol,Int}, parameters::OrderedDict{Symbol,Int})
+# Generate the Jacobian evaluation function as a single expression.
+# zeroout_jac kwarg controls whether to first fill the matrix with zeros and then only fill in non-zero entries
+# vs looping through and putting expressions for entries that are always zero.
+# This is needed for sufficiently large dense matrices to allow compilation. (e.g. 1000x1000)
+function calculate_jac(symjac::Matrix{ExprValues}, reactants::OrderedDict{Symbol,Int}, parameters::OrderedDict{Symbol,Int}, zeroout_jac=false)
     func_body = Expr(:block)
-    @inbounds for j = 1:size(symjac)[2], i = 1:size(symjac)[1]
-        ex =  (symjac[i,j] isa Number) ? symjac[i,j] : recursive_replace!(symjac[i,j],(reactants,:internal___var___u), (parameters, :internal___var___p))
-        push!(func_body.args, :(internal___var___J[$i,$j] = $(ex)))
+    zeroout_jac && push!(func_body.args, :(fill!(internal___var___J, zero(eltype(internal___var___J)))))
+    @inbounds for j = 1:size(symjac)[2], i = 1:size(symjac)[1]        
+        if symjac[i,j] isa Number
+            ex = symjac[i,j]
+            (!zeroout_jac || !iszero(ex)) && push!(func_body.args, :(internal___var___J[$i,$j] = $(ex)))
```

Comment:
This maybe a bit subtle. The idea is that for large enough networks, like the BCR network, we can't generate a function with one line for each Jacobian entry (we would have to split the function into subfunctions). However, we would like a dense Jacobian as the matrix is not so large that they don't work well. (As I mentioned, once compiled it gives a faster solver with KenCarp4 than using finite differences.)

If we first zero out the matrix in the generated function we only need to then fill in non-zero Jacobian entries, which are usually sparse for larger systems. Setting `zeroout_jac=true` takes the latter approach. The default is `false` to keep the current DSL Jacobian behavior (where even zero entries in the Jacobian get a line in the generated function.)